### ISSUE TITLE
Add syntactically-not-free SynNF to set.mm/iset.mm?

### ISSUE THREAD

**Issue by david-a-wheeler** [metamath/set.mm#1183]
I'm thinking it would be useful to add a syntactically-not-free construct to set.mm/iset.mm.

As you know, it's very common in textbooks to make statements like "where x is not free in QQQ", where "not free" means the variable is not present or is bound by a quantifier.  Both set.mm and iset.mm have `df-nf` that defines `F/` as the not-free operation, specifically as `|- ( F/ x ph <-> A. x ( ph -> A. x ph ) ) $.`  But `F/` is not quite the same as textbook definitions, for example, `F/ x x = x` is true.

Issue #1173 discusses an experimental approach to formalize bounded formulas in iset.mm.  It seems to me this same approach could be used to formalize "syntactically not free" in both set.mm and iset.mm.

Basically, something like `SynNF x ph` would be true iff x is syntactically not free in ph.  This is not exactly the same as `F/`, because expressions such as `SynNF x x = x` would be false, while `F/ x x = x` is true. So A. x ( SynNF x ph -> F/ x ph ) but not the other way.

Doing this would require adding some $a statements, but I think these are more like specialized definitions than true axioms since they don't add to the set of things that can be proved, they instead provide an implementation of a common notation.

Is that reasonable? Madness?

**Comment by david-a-wheeler** [comment]
I haven't thought too hard about what axioms would be needed to implement `SynNF x ph`. But looking at the discussion in #1173, especially @digama0 , I think it'd look something like these axioms in classical logic (not sure what it'd look like in intuitionistic logic):

~~~~
|- SynNF x A. x ph
|- ( ( SynNF x ph /\ SynFS x ps ) -> SynNF x ( ph -> ps ) )
|- SynNF x ph <-> SunNF x -. ph
( ph <-> ps ) |- ( SynNF x ph -> SynNF x ps )
~~~~

Mario described how to implement Bdd and included `$d x y |- Bdd ph -> Bdd A. x e. y ph`; I'm not sure if that should be ported over to SynFS (or how).

The idea is that you could then prove that certain other expressions were also SynNF.

I think you'd need a SynNF_ for classes, analogous to F/_.  I think that could defined in a way similar to how `F/_` is already defined in `df-nfc` and would be a very conventional definition. First try (though I'm not sure "y" is handled correctly here):

~~~~
|- ( SynNF_ x A <-> A. y SynNF x y e. A )
~~~~

**Comment by david-a-wheeler** [comment]
Hmm, you'd also need something like this so that "variable not in expression" worked:

`$d x ph |- SynNF x ph`

**Comment by david-a-wheeler** [comment]
Hmm, missing something. If we accept:

`( ph <-> ps ) |- ( SynNF x ph -> SynNF x ps )`

That doesn't work.  `( x = x <-> T. )` but should have different SynNF values.

It's past midnight my time, too late to be trying this :-).

**Comment by digama0** [comment]
> `( ph <-> ps ) |- ( SynNF x ph -> SynNF x ps )`

If you add this axiom, it will make `SynNF` the same as `F/`. In fact, this is an equivalent way to characterize `F/`: it holds of `ph` if `ph` is equivalent to a formula in which `x` is syntactically not free.

> Basically, something like SynNF x ph would be true iff x is syntactically not free in ph. This is not exactly the same as F/, because expressions such as SynNF x x = x would be false, while F/ x x = x is true. So A. x ( SynNF x ph -> F/ x ph ) but not the other way.

You have to be careful about the difference between true and provable here. You can axiomatize `SynNF` such that metatheoretically, `|- SynNF x ph` is provable iff `x` is syntactically not free in `ph` (when `ph` contains no wff metavariables), but that doesn't mean that you can ever prove `|- -. SynNF x ph`, and in particular, just because "`SynNF x x = x` would be false" doesn't mean you can prove that in metamath's logic.

To axiomatize the "negative" part of `SynNF` would be much harder, and I'm not even sure it's possible. You would need to reify in the logic all possible ways to construct wffs (a.k.a. "proof by induction on wffs"), which runs up against some impossibility theorems because `SynNF` itself is a new wff constructor, and so you have a circularity that quite likely causes inconsistency.

**Comment by digama0** [comment]
> @digama0 said:
>
> > The axioms are conservative, because you can model SynNF x ph with F/ x ph.
>
> They aren't exactly the same. `SynNF x x = x` would be false, while `F/ x x = x` is true. I realize you're using a technical meaning for "model", but I mean "produces the same result". So `A. x ( SynNF x ph -> F/ x ph )` but not the other way.

Right, they aren't the same, but the logic can't tell the difference. This is actually why `F/` has been as effective as it is, because the only proper predicate that acts like `SynNF` is `F/`. `SynNF` is not a predicate because it's not closed under equivalence. If you take all the axioms for `SynNF` and replace them with `F/`, the axioms are all provable. In fact, even if you replace the axioms for `SynNF x ph` with simply `SynNF x ph <-> T.`, the axioms are all provable. This has to do with what I said above about not characterizing the negative part of `SynNF`.

Really, the only way `SynNF` becomes interesting is if it appears somewhere in negative position, meaning that you have some axiom that assumes `SynNF` as a hypothesis. In FOL usually this would be some kind of forall introduction rule or similar which requires a side condition about the context. Because the conclusion of this rule will involve something that is closed under equivalence, we are licensed to close `SynNF` under equivalence as well, and `F/` is the result.

> I certainly appreciate that @nmegill wants to minimize the number of axioms; makes sense to me too. But this wouldn't really an "axiom" in the usual sense, instead, it's a *definition* that happens to use $a to represent it.  Providing a mechanism to describe "syntactically not free" doesn't change what the underlying system can do.  In addition, "syntactically not free" is a pretty common case in mathematical notation.

This is a *really* slippery slope. There are many instances of our trying to explain away axioms by this method, but it doesn't really work. Calling it a definition instead of an axiom is just a linguistic trick that may help to make it feel more justified, but it's really not a definition, and it requires just as much care as any other axiom addition.

Let's look at the usual properties satisfied by definitions.
* Is it eliminable? No, it's provable in certain circumstances and unknown in others.
* Is it a definitional extension (that is, the models of the extension uniquely determine the new predicate in terms of the others)? No; because the negative part is unconstrained, even `T.` is a model for `SynNF`, and syntactically not free is another model, so it's not uniquely determined.
* Is it a conservative extension? Yes, because if you replace `SynNF` with `T.` everywhere (or `F/` if you prefer) then you have a proof in the original system.
* More pragmatically, this obviously does not satisfy the requirements imposed by the MMJ2 definition checker.

**Comment by benjub** [comment]
@davidawheeler : In the first comment of Issue #1173, I refer to two
earlier threads and to the head comment of my iset.mm mathbox, which I
advise you to read.  You'll see that, in defining bounded formulas, we
decided to opt for "semantically bounded" and not "syntactically bounded"
because it is much more convenient when working with defined symbols.  If
you don't do that, you'll have to either give up using defined symbols,
resulting in overly large formulas, or to add an axiom corresponding to
each definition (e.g.: you define df-or? then you need to add ax-snfor! you
define df-an? then you need to add ax-snfan! and similarly with bi, 3or,
3an, nan, xor, tru, fal, etc.).  Therefore, I think that SynNF would not
add much, except for complications, to F/, which does perfectly the job (at
least for the current purposes of set.mm).  Mario's argument shows in a
precise way that F/ is the "good replacement" for SynNF, being closed under
equivalence.

**Comment by DavidAWheeler** [comment]
hi, fren

**Comment by david-a-wheeler** [comment]
@benjub - be careful, I am @david-a-wheeler .  The account @DavidAWheeler is a different person :-).

**Comment by DavidAWheeler** [comment]
we are all DAW this fine day

**Comment by david-a-wheeler** [comment]
Makes sense. Closing.

**Comment by Nikolaj-K** [comment]
Hello guys and @jkingdon , I have an interest in CZF and have been looking into your work on the metamath site. Since it look still quite active and as I have only seen one reference to a 2001 text by Aczel on the green Intuitionistic page, I thought I point to this 2010 draft that's more than twice as long as the 2001 "notes" pdf. 
https://www1.maths.leeds.ac.uk/~rathjen/book.pdf
I also have collected a few more relevant papers and slides at the start here
https://gist.github.com/Nikolaj-K/951d44b393a92acc51f915fcb0e31cc2

**Comment by jkingdon** [comment]
> Hello guys and @jkingdon , I have an interest in CZF and have been looking into your work on the metamath site. Since it look still quite active and as I have only seen one reference to a 2001 text by Aczel on the green Intuitionistic page, I thought I point to this 2010 draft that's more than twice as long as the 2001 "notes" pdf.
> https://www1.maths.leeds.ac.uk/~rathjen/book.pdf
> I also have collected a few more relevant papers and slides at the start here
> https://gist.github.com/Nikolaj-K/951d44b393a92acc51f915fcb0e31cc2

Hey, welcome to metamath and hope you find something you are interested in working on.

Most of the work on CZF has been done by @benjub and can be found at http://us.metamath.org/ileuni/ax-bd0.html , the section header right above it, and following axioms/theorems.

I'm sure there is more to do. Every time I use power set, I think of CZF, but my own work has mostly been with IZF. Let us know if you have some CZF proofs and need to know where to put them, or there's anything else we can advise on.

**Comment by Nikolaj-K** [comment]
Okay, thanks. 
Since the [definitions list](http://us.metamath.org/ileuni/mmdefinitions.html) for the Intuitionistic section seems to have all IZF definitions/axioms and only the bottom has a few CZF specific definitions, it's a bit hard for me to keep the two separate axioms apart and, looking at a theorem, know what the context theory is. 
Since CZF, IZF, Z, ZF, ZFC all used the same language, I suppose there's also a bit redundancy in things proven.

**Comment by digama0** [comment]
@Nikolaj-K I would hope that the intuitionistic logic database can serve as a playground for all the intuitionistic and constructive axiomatizations of predicate calculus and set theory, so I wouldn't expect there to be much duplication beyond the basic duplication wrt set.mm. That is, I wouldn't expect separate CZF and IZF databases, because they are compatible enough that they can share most of the common theory in the intersection of their axiomatizations, each with perhaps a few extra theorems of their own (and maybe some that can be proven with either system but not the intersection? Not sure how common this is).

In theory we could do the same thing with set.mm, but I think the judgment was that this would involve changing the statements of too many theorems, in many cases resulting in a weaker statement that is more work to apply.

**Comment by benjub** [comment]
As a side remark, in my development of CST, I used (non-systematically) Aczel and Rathjen's CST Book draft (see head comment of the section "Bounded formulas") and I took care of axiom use in the proofs (the web page for each statement has a line "This theorem was proved from axioms" where you can get this information). In particular, one can see which results are proved in ECST and BCST.
